README.md
# hrtime
Human-Readable Time, stylized as `hrtime`, is a thin Rust library which
converts seconds into either a colon-seperated time string and vice versa,
or into the raw hour, minute, and second values.

## Why?
Seemed like a simple library to make as I am learning Rust and thought it
may prove useful to someone. As you can see by the versioning (0.2.0) it
is currently in very early stages.

## How?

### From seconds
This crate only contains four functions, two "from" functions: `from_sec`
and `from_sec_padded`, and two "to" functions: `to_sec` and `to_time`.
The two "from" functions will convert a given `u64` into a colon-separated
time string, with `from_sec_padded` specifically introducing leading
zeroes to reach the format `"00:00:00"` (HH:MM:SS). An example using
`from_sec`:
```rust
let secs = 234;
println!("{secs} seconds is {}", hrtime::from_sec(secs));
```
Will print `"234 seconds is 3:54"`,

and the same example but using `from_sec_padded`:
```rust
let secs = 234;
println!("{secs} seconds is {}", hrtime::from_sec_padded(secs));
```
Will print `"234 seconds is 00:03:54"`!

### To seconds
The first "to" function is `to_sec`, which takes in a `time` string as an
argument and attempts to convert it into the amount of seconds it
represents. This string has some requirements to meet in order to work
though, like having the values separated by colons (ex: `"1:38"` for a
minute and 38 seconds) and there being no more than three colons (ex:
`"1:23:14:38"` will panic). `to_sec` may be used like so:
```rust
let time = "10:50";
let secs = hrtime::to_sec(time);
println!("{time} is {secs} seconds");
```
Which prints `"10:50 is 650 seconds"`!

### To time
The second "to" function is `to_time`, which returns the raw hour, minute,
and second values represented by the given `u64` respectively. This is
returned as a tuple containing a `u64` (hour), and two `u8`s (minute and
seconds) more information can be seen in its documentation. `to_time` may
be used like so:
```rust
let seconds = 650;
let (hrs, min, sec) = hrtime::to_time(seconds);
println!("{seconds} seconds is {hrs}h{min}m{sec}s");
```
Which prints `"650 seconds is 0h10m50s"`!

### More
More examples can be seen in the
[examples](https://github.com/marcelohdez/hrtime/tree/master/examples)
folder.

## License
`hrtime` is licensed under the MIT license, for more information please
read the
[LICENSE file](https://github.com/marcelohdez/hrtime/blob/master/LICENSE).


CHANGELOG.md
# `hrtime` Changelog

## 0.2.0
* New: `to_time` function
    * Returns the raw hour, minute, and second values from the given
    seconds `u64` respectively.

src/tests.rs
#[cfg(test)]
mod tests {
    // ======= from_sec =======
    #[test]
    fn test_from_sec_sec() {
        let result = crate::from_sec(43);
        assert_eq!(&result, "43");
    }

    #[test]
    fn test_from_sec_min() {
        let result = crate::from_sec(100);
        assert_eq!(&result, "1:40");
    }

    #[test]
    fn test_from_sec_hr() {
        let result = crate::from_sec(3671);
        assert_eq!(&result, "1:01:11");
    }

    // ======= from_sec_padded =======
    #[test]
    fn test_from_sec_padded_sec() {
        let result = crate::from_sec_padded(55);
        assert_eq!(&result, "00:00:55");
    }

    #[test]
    fn test_from_sec_padded_min() {
        let result = crate::from_sec_padded(892);
        assert_eq!(&result, "00:14:52");
    }

    #[test]
    fn test_from_sec_padded_hrs() {
        let result = crate::from_sec_padded(3999);
        assert_eq!(&result, "01:06:39");
    }

    // ======= to_sec =======
    #[test]
    fn test_to_sec_sec() {
        let time = "2";
        let result = crate::to_sec(time);
        assert_eq!(result, 2);
    }

    #[test]
    fn test_to_sec_min() {
        let time = "37:18";
        let result = crate::to_sec(time);
        assert_eq!(result, 2238);
    }

    #[test]
    fn test_to_sec_hrs() {
        let time = "8:29:17";
        let result = crate::to_sec(time);
        assert_eq!(result, 30557);
    }

    // ======= to_time =======
    #[test]
    fn test_to_time_sec() {
        let seconds = 40;
        let result = crate::to_time(seconds);
        assert_eq!(result, (0, 0, 40));
    }

    #[test]
    fn test_to_time_min() {
        let seconds = 378; // 6 minutes, 18 seconds
        let result = crate::to_time(seconds);
        assert_eq!(result, (0, 6, 18));
    }

    #[test]
    fn test_to_time_hr() {
        let seconds = 3710; // 1 hour, 1 minute, 50 seconds
        let result = crate::to_time(seconds);
        assert_eq!(result, (1, 1, 50));
    }
}


src/lib.rs
mod tests;

/// Convert the given `secs` u64 into a readable time format without any
/// leading zeroes or colons, ex: `318` -> `"5:18"`.
/// 
/// If you would like leading zeroes on the output (ex: `1200` -> `"00:20:00"`)
/// you may use `from_sec`'s sister function, [`from_sec_padded`].
/// 
/// # Example:
/// ```
/// // Prints out "1200 seconds is 20:00"!
/// let mut sec = 1200;
/// println!("{sec} seconds is {},", hrtime::from_sec(sec));
/// // Passes:
/// assert_eq!("1:00", hrtime::from_sec(60));
/// ```
pub fn from_sec(secs: u64) -> String {
    let (hrs, min, sec) = to_time(secs);

    // 0>2 pads the number with 0s to the left if less than 2 digits wide
    if hrs > 0 { // If there are hours to show:
        format!("{hrs}:{min:0>2}:{sec:0>2}")
    } else if min > 0 { // Else if there are minutes to show:
        format!("{min}:{sec:0>2}")
    } else { // If there are only seconds to show:
        format!("{sec}")
    }
}

/// Much like [`from_sec`], this function will convert the given `secs` u64
/// into a readable time format. But unlike [`from_sec`], this function
/// will introduce leading 0's until it reaches the format `"00:00:00"`
/// 
/// # Example
/// ```
/// // Prints "28309 seconds with padding is 07:51:49"!
/// let sec = 28309;
/// println!("{sec} seconds with padding is {}", hrtime::from_sec_padded(sec));
/// // Passes:
/// assert_eq!("01:00:00", hrtime::from_sec_padded(3600));
/// ```
pub fn from_sec_padded(secs: u64) -> String {
    let (hrs, min, sec) = to_time(secs);
    // 0>2 pads the number with 0s to the left if less than 2 digits wide
    format!("{hrs:0>2}:{min:0>2}:{sec:0>2}")
}

/// Will attempt to convert the given `time` string into a u64 of seconds.
/// `time` must have the hours, minutes, and seconds separated by colons
/// respectively in order for this function to work correctly.
/// Ex: `"1:00"`, `"1:00:00"`, or `"2:38"`.
/// 
/// # Panics
/// Will panic if the given `time` string contains more than three colons
/// (ex: `"1:23:40:14"`), or if any of the time values before or after the
/// colons cannot be parsed as u64 integers (ex: `":23"` or `"12:"`).
/// 
/// # Examples
/// ## Hours, minutes, and seconds:
/// ```
/// // Prints "11:23:08 is 40988 seconds"!
/// let time = "11:23:08";
/// let sec = hrtime::to_sec(time);
/// println!("{time} is {sec} seconds");
/// // Passes:
/// assert_eq!(120, hrtime::to_sec("2:00"));
/// ```
/// 
/// ## Just seconds:
/// ```
/// // Prints "12 is 12 seconds"!
/// let time = "12";
/// let sec = hrtime::to_sec(time);
/// println!("{time} is {sec} seconds");
/// // Passes:
/// assert_eq!(48, hrtime::to_sec("48"));
/// ```
pub fn to_sec(time: &str) -> u64 {
    let mut sec = 0;
    let minpos = time.rfind(':'); // Position of minute colon (rightmost)

    // Check if minpos contains the index `i` of a colon:
    if let Some(i) = minpos {
        let hrpos = time[0..i].rfind(':'); // Position of hour colon (next colon from minute one)
        let mut start = 0;

        if let Some(pos) = hrpos { // If there is an hour colon:
            // Parse the value from the start to the hour colon
            sec += time[0..pos]
                    .parse::<u64>()
                    .unwrap()
                    * 60 * 60; // * 60 for min and * 60 for hr

            start = pos + 1;
        }

        // Add the parsed minute value
        sec += time[start..i]
                .parse::<u64>()
                .unwrap()
                * 60; // Convert into mins

        sec += time[i + 1..time.len()].parse::<u64>().unwrap(); // Convert text after this colon into seconds
    } else { // If there is no colon, treat text as only seconds:
        sec += time.parse::<u64>().unwrap();
    }

    sec
}

/// Returns the hours, minutes, and seconds respectively that are
/// represented in the given seconds `u64`.
/// 
/// ## Why are the minute and second values `u8`s?
/// Since both of these values will always return as <= 59, any larger
/// integer size is redundant. Therefore, if your program needs a
/// different integer range (outside of 0-255) remember to cast these
/// values as such!
/// 
/// # Example
/// ```
/// let seconds = 3661; // One hour, one minute, and one second.
/// let (hrs, min, sec) = hrtime::to_time(seconds);
/// // Prints "3661 seconds is 1h1m1s!"
/// println!("{seconds} seconds is {hrs}h{min}m{sec}s!");
/// 
/// // Passes:
/// assert_eq!((0, 2, 15), hrtime::to_time(135));
/// ```
pub fn to_time(secs: u64) -> (u64, u8, u8) {
    let sec = (secs % 60) as u8;
    let min = ((secs / 60) % 60) as u8;
    let hrs = secs / 60 / 60;

    (hrs, min, sec)
}
